C7Engine/MapUnitExtensions.cs

```diff
@@ -102,53 +97,111 @@ public static void RollToPromote(this MapUnit unit, MapUnit opponent, bool waitF
+		double attackerStrength = attacker.unitType.attack  * StrengthBonus.ListToMultiplier(attackBonuses),
 		       defenderStrength = defender.unitType.defense * StrengthBonus.ListToMultiplier(defenseBonuses);
 
-		Console.WriteLine($"Combat log: {unit.unitType.name} ({attackerStrength}) attacking {defender.unitType.name} ({defenderStrength})");
-		Console.WriteLine($"\tAttacker: {unit.unitType.name}, base strength {unit.unitType.attack}");
+		Console.WriteLine($"Combat log: {attacker.unitType.name} ({attackerStrength}) attacking {defender.unitType.name} ({defenderStrength})");
+		Console.WriteLine($"\tAttacker: {attacker.unitType.name}, base strength {attacker.unitType.BaseStrength(CombatRole.Attack)}");
 		foreach (StrengthBonus bonus in attackBonuses)
 			Console.WriteLine($"\t\t+{100.0*bonus.amount}%\t{bonus.description}");
-		Console.WriteLine($"\tDefender: {defender.unitType.name}, base strength {defender.unitType.defense}");
+		Console.WriteLine($"\tDefender: {defender.unitType.name}, base strength {defender.unitType.BaseStrength(CombatRole.Defense)}");
 		foreach (StrengthBonus bonus in defenseBonuses)
 			Console.WriteLine($"\t\t+{100.0*bonus.amount}%\t{bonus.description}");
 
+		CombatResult result = CombatResult.Impossible;
+
 		double attackerOdds = attackerStrength / (attackerStrength + defenderStrength);
 		if (Double.IsNaN(attackerOdds))
-			return false;
+			return result;
+
+		// Defensive bombard
+		MapUnit defensiveBombarder = MapUnit.NONE;
+		double defensiveBombarderStrength = 0.0;
+		foreach (MapUnit candidate in defender.location.unitsOnTile.Where(u => u != defender && u.defensiveBombardsRemaining > 0)) {
+			double strength = candidate.StrengthVersus(attacker, CombatRole.DefensiveBombard, attacker.facingDirection.reversed());
+			if (strength > defensiveBombarderStrength) {
+				defensiveBombarder = candidate;
+				defensiveBombarderStrength = strength;
+			}
+		}
+		if (defensiveBombarder != MapUnit.NONE &&
+		    attacker.hitPointsRemaining > 1) { // TODO: Can defensive bombard kill? What does the original game do in this case?
+			var dBOriginalDirection = defensiveBombarder.facingDirection;
+			defensiveBombarder.facingDirection = defender.facingDirection;
+
+			defensiveBombarder.animate(MapUnit.AnimatedAction.ATTACK1, true);
+
+			// dADB = defense Against Defensive Bombard
+			double dADB = attacker.StrengthVersus(defensiveBombarder, CombatRole.DefensiveBombardDefense, defensiveBombarder.facingDirection);
+			if (EngineStorage.gameData.rng.NextDouble() < defensiveBombarderStrength / (defensiveBombarderStrength + dADB))
+				attacker.hitPointsRemaining -= 1;
+
+			defensiveBombarder.defensiveBombardsRemaining -= 1;
+			defensiveBombarder.facingDirection = dBOriginalDirection;
+		}
+
+		// TODO: Check these rules are accurate and that we're not missing any.
+		bool defenderEligibleToRetreat = defender.hitPointsRemaining > 1 && ! defender.location.HasCity;
```

Comment:
There's at least one other case, which is that there has to be a valid tile to retreat to.  At one level, this means the unit can't be up against the sea, but it _also_ means the unit hasn't been fully flanked.  I'd have to check to be 100% sure of the rules, but my understanding has been that the tiles "behind" the unit after it is rotated to face its attack must not be occupied.  But all of them not be occupied, just one, depends on ZOC?  That's where it's tricky.

I propose writing up an issue for the check-for-a-valid-tile-to-retreat-to and referencing that issue in the TODO, since it could be tricky and probably belongs in Combat Part III.

Reply:
Ah, I see that this is at least somewhat covered on line 169.  We probably do still need to have a test to document the original game's rules on the full extent of this.

Reply:
Okay, it's at least mostly covered.  Only question remaining is whether a unit adjacent to both the unit and the tile it wants to retreat to prevents that retreat.  Well, and I guess if it can retreat at a "diagonal" if the tile behind it is not of a valid terrain type, but the tile behind-and-to-the-side is of a valid terrain type.

Reply:
I opened issue #274 for the matter of perfecting defender retreat behavior, and updated that comment & a couple of others in light of your remarks about the combat rules. I hadn't even considered zone of control until you mentioned it, that's one of those little rules I always forget about until I get hit by it.